src/GitVersionCore/Core/Abstractions/Git/IGitRepository.cs

```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Collections.Generic;
+
+namespace GitVersion
+{
+    public interface IGitRepository : IDisposable
+    {
+        string Path { get; }
+        string WorkingDirectory { get; }
+        bool IsHeadDetached { get; }
+        IBranch Head { get; }
+        ITagCollection Tags { get; }
+        IReferenceCollection Refs { get; }
+        IBranchCollection Branches { get; }
+        ICommitCollection Commits { get; }
+
+        IGitRepository CreateNew(string gitRootPath);
+
+        int GetNumberOfUncommittedChanges();
+        string ShortenObjectId(ICommit commit);
+
+        bool GitRepoHasMatchingRemote(string targetUrl);
+        void CleanupDuplicateOrigin(string gitRootPath, string remoteName);
+        bool GetMatchingCommitBranch(ICommit baseVersionSource, IBranch branch, ICommit firstMatchingCommit);
+        IRemote EnsureOnlyOneRemoteIsDefined();
+        ICommit FindMergeBase(ICommit commit, ICommit otherCommit);
+        ICommit GetBaseVersionSource(ICommit currentBranchTip);
+        ICommit GetForwardMerge(ICommit commitToFindCommonBase, ICommit findMergeBase);
+
+        IEnumerable<ICommit> GetCommitsReacheableFrom(ICommit commit, IBranch branch);
+        IEnumerable<ICommit> GetMergeBaseCommits(ICommit mergeCommit, ICommit mergedHead, ICommit findMergeBase);
+        IEnumerable<ICommit> GetMainlineCommitLog(ICommit baseVersionSource, ICommit mainlineTip);
+        IEnumerable<ICommit> GetCommitsReacheableFromHead(ICommit headCommit);
+        IEnumerable<ICommit> GetCommitLog(ICommit baseVersionSource, ICommit currentCommit);
+
+        void Checkout(string commitOrBranchSpec);
+        void Checkout(IBranch branch);
+        void Fetch(string remote, IEnumerable<string> refSpecs, AuthenticationInfo auth, string logMessage);
+        void CreateBranchForPullRequestBranch(AuthenticationInfo auth);
+        string Clone(string sourceUrl, string workdirPath, AuthenticationInfo auth);
```

Comment:
I'm not sure how some of these methods are used, but `Clone` returning `string` seems a bit impractical to me. Should we revise and modify the return values of some of these methods and perhaps debate their very existence as well?

Reply:
As far as I remember all these methods are used by GitPrepare, as this is the only part that does changes to a repository. The rest of the GitVersionCore is just reading the git data. For now I would keep it like this, then in v6 when we have the normalize command we'll move that part there.